java/sage/nio/RemoteFileChannel.java

```diff
@@ -0,0 +1,703 @@
+    int bytes = read(singleByte);
+    singleByte.flip();
+
+    if (bytes == -1)
+      throw new java.io.EOFException();
+
+    return singleByte.get() & 0xFF;
+  }
+
+  @Override
+  public int read(ByteBuffer dst) throws IOException
+  {
+    int returnValue = read(dst, remoteOffset);
+
+    if (returnValue == -1)
+      return returnValue;
+
+    remoteOffset += returnValue;
+    return returnValue;
+  }
+
+  @Override
+  public synchronized long read(ByteBuffer[] dsts, int offset, int length) throws IOException
+  {
+    if (length == 0)
+      return 0;
+
+    long returnValue = 0;
+    long readBytes;
+    long totalLength = 0;
+
+    int remains;
+    boolean bufferAvailable = false;
+    int limit = offset + length;
+    for (int i = offset; i < limit; i++)
+    {
+      remains = dsts[i].remaining();
+
+      // Since we are iterating through the buffers either way, we might as well find the first one
+      // that actually has space available.
+      if (remains > 0)
+      {
+        bufferAvailable = true;
+        totalLength += remains;
+      }
+      else if (!bufferAvailable)
+      {
+        offset += 1;
+      }
+
+      // Don't continue if it's not going to matter.
+      if (totalLength > maxRemoteSize && totalLength > size())
+        break;
+    }
+
+    // All of the buffers do not have space available.
+    if (offset == length)
+      return 0;
+
+    totalLength = getMaxRead(remoteOffset, totalLength);
+    readBytes = startRead(remoteOffset, totalLength, dsts[offset++], true);
+    if (readBytes == -1)
+      return -1;
+
+    returnValue += readBytes;
+
+    // The first buffer is already fully used.
+    for (int i = offset; i < length; i++)
+    {
+      while (dsts[i].hasRemaining())
+      {
+        readBytes = readMore(dsts[i]);
+
+        if (readBytes == -1)
+          break;
+
+        returnValue += readBytes;
+      }
+    }
+
+    remoteOffset += returnValue;
+    return returnValue;
+  }
+
+  @Override
+  public int write(ByteBuffer src) throws IOException
+  {
+    if (readonly)
+      throw new IOException("Channel is read only.");
+
+    int length = write(src, remoteOffset);
+    remoteOffset += length;
+    return length;
+  }
+
+  @Override
+  public synchronized long write(ByteBuffer[] srcs, int offset, int length) throws IOException
+  {
+    if (readonly)
+      throw new IOException("Channel is read only.");
+
+    if (length == 0)
+      return 0;
+
+    long returnValue;
+    long totalLength = 0;
+
+    int remains;
+    boolean bufferAvailable = false;
+    int limit = offset + length;
+    for (int i = offset; i < limit; i++)
+    {
+      remains = srcs[i].remaining();
+
+      // Since we are iterating through the buffers either way, find the first one that actually has
+      // data available.
+      if (remains > 0)
+      {
+        bufferAvailable = true;
+        totalLength += remains;
+      }
+      else if (!bufferAvailable)
+      {
+        offset += 1;
```

Comment:
Changed things slightly, but according to several different scenarios I ran through the debugger, this code is working exactly as expected.